File: benchmarks/asyncio/bench_io.go
// I/O-Bound Benchmark: Concurrent Sleep (Go goroutines)
package main

import (
	"fmt"
	"sync"
	"time"
)

const (
	NUM_TASKS = 10000
	SLEEP_MS  = 1
)

func worker(taskID int) int {
	time.Sleep(SLEEP_MS * time.Millisecond)
	return taskID
}

func main() {
	start := time.Now()

	results := make(chan int, NUM_TASKS)
	var wg sync.WaitGroup

	for i := 0; i < NUM_TASKS; i++ {
		wg.Add(1)
		go func(taskID int) {
			defer wg.Done()
			results <- worker(taskID)
		}(i)
	}

	go func() {
		wg.Wait()
		close(results)
	}()

	var total int = 0
	for result := range results {
		total += result
	}

	elapsed := time.Since(start)

	fmt.Println("Benchmark: I/O-bound")
	fmt.Printf("Tasks: %d\n", NUM_TASKS)
	fmt.Printf("Sleep per task: %dms\n", SLEEP_MS)
	fmt.Printf("Total result: %d\n", total)
	fmt.Printf("Time: %.2fms\n", float64(elapsed.Nanoseconds())/1e6)
	fmt.Printf("Tasks/sec: %.0f\n", float64(NUM_TASKS)/(float64(elapsed.Nanoseconds())/1e9))
	fmt.Printf("Sequential would be: %dms\n", NUM_TASKS*SLEEP_MS)
	fmt.Printf("Concurrency factor: %.0fx\n", float64(NUM_TASKS*SLEEP_MS)/(float64(elapsed.Nanoseconds())/1e6))
}

File: benchmarks/asyncio/bench_io.py
"""
I/O-Bound Benchmark: Concurrent Network Simulation

Tests concurrent I/O handling performance:
- Spawns N worker tasks
- Each worker simulates network latency (sleep)
- Collects results via gather

Best for: event loop / netpoller
"""
import asyncio
import time

NUM_TASKS = 10000
SLEEP_MS = 100  # 100ms simulated I/O latency

async def worker(task_id: int) -> int:
    """I/O-bound worker (simulated network call)"""
    await asyncio.sleep(SLEEP_MS / 1000)
    return task_id

async def main():
    start = time.perf_counter()

    tasks = [worker(i) for i in range(NUM_TASKS)]
    results = await asyncio.gather(*tasks)

    elapsed = time.perf_counter() - start
    total = sum(results)

    print(f"Benchmark: I/O-bound")
    print(f"Tasks: {NUM_TASKS}")
    print(f"Sleep per task: {SLEEP_MS}ms")
    print(f"Total result: {total}")
    print(f"Time: {elapsed*1000:.2f}ms")
    print(f"Tasks/sec: {NUM_TASKS/elapsed:.0f}")
    print(f"Sequential would be: {NUM_TASKS * SLEEP_MS}ms")
    print(f"Concurrency factor: {(NUM_TASKS * SLEEP_MS) / (elapsed * 1000):.0f}x")

if __name__ == "__main__":
    asyncio.run(main())
